props/websubmit/src/main.rs

```diff
@@ -85,16 +93,108 @@ impl DeletionProp {
+}
+
+/// Storing data in the database must be associated to a user. This is
+/// necessary for e.g. the deletion to work.
+pub struct ScopedStorageProp {
+    cx: Arc<PolicyContext>,
+}
+
+impl ScopedStorageProp {
+    pub fn new(cx: Arc<PolicyContext>) -> Self {
+        ScopedStorageProp { cx }
+    }
+
+    pub fn check(self) {
+        for c_id in self.cx.desc().controllers.keys() {
+            let scopes = self
+                .cx
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|node| self.cx.has_marker(marker!(scopes), *node))
+                .collect::<Vec<_>>();
+            let stores = self
+                .cx
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|node| self.cx.has_marker(marker!(stores), *node))
+                .collect::<Vec<_>>();
+            let mut sensitives = self
+                .cx
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|node| self.cx.has_marker(marker!(sensitive), *node));
+
+            let controller_valid = sensitives.all(|sens| {
+                stores.iter().all(|&store| {
+                    // sensitive flows to store implies some scope flows to store callsite
+                    !(self
+                        .cx
+                        .flows_to(sens, store, paralegal_policy::EdgeType::Data))
+                        || store.typ.as_call_site().is_some_and(|cs| {
+                            // The sink that scope flows to may be another CallArgument attached to the store's CallSite, it doesn't need to be store itself.
+                            let found_scope = scopes.iter().any(|scope| {
+                                self.cx.flows_to(
+                                    *scope,
+                                    Node {
+                                        ctrl_id: store.ctrl_id,
+                                        typ: cs.into(),
+                                    },
+                                    paralegal_policy::EdgeType::Data,
+                                )
+                            });
+                            assert_error!(
+                                self.cx,
+                                found_scope,
+                                format!(
+                                    "Stored sensitive isn't scoped. sensitive {} stored here: {}",
+                                    self.cx.describe_node(sens),
+                                    self.cx.describe_node(store)
+                                )
+                            );
+                            found_scope
+                        })
+                })
+            });
+
+            assert_error!(
+                self.cx,
+                controller_valid,
+                format!(
+                    "Violation detected for controller: {}",
+                    self.cx.describe_def(*c_id)
+                ),
+            );
+        }
+    }
+}
+
+fn sc_policy(ctx: Arc<Context>) -> Result<()> {
```

Comment:
Minor nit: I would name these functions `run_sc_policy` and `run_del_policy` to distinguish them from their structs. And I would use "Scoped Storage" and "Deletion" (without the "Policy") for the argument to `named_policy` because the error messages add it automatically.

Reply:
done